helm,test: Add standalone L4LB XDP tests in a form of Github Action

brb opened the pull request:
This PR introduces a new GH action called `Cilium L4LB XDP` which is responsible for running the standalone LB tests.

The action starts a Fedora VM with vagrant. We do that because we need to run Kind on cgroupv2-only machine (otherwise, bpf_sock which is required by the LB health check is not guaranteed to work). Unfortunately, GH Action does not support any runner with cgroupv2-only. So instead we run Fedora 34 which has cgroupv1 disabled on the MacOS runner which supports nested virtualisation.

For now the test issues 10 requests to LB VIP from the Fedora VM. See `test.sh` for more details.

In addition, the PR adds the `loadBalancer.standalone` and friend options to enable the standalone L4LB.

Reviewable per commit.

brb committed:
test: Add GH Action for Cilium L4LB XDP tests

This commit introduces a new GH action called "Cilium L4LB XDP" which is
responsible for running the standalone LB tests.

The action starts a Fedora VM with vagrant. We do that because we need
to run Kind on cgroupv2-only machine (otherwise, bpf_sock which is
required by the LB health check is not guaranteed to work).
Unfortunately, GH Action does not support any runner with cgroupv2-only.
So instead we run Fedora 34 which has cgroupv1 disabled on the MacOS
runner which supports nested virtualisation.

For now the test issues 10 requests to LB VIP from the Fedora VM. See
test.sh for more details.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on test/l4lb-xdp/test.sh:
```diff
@@ -0,0 +1,70 @@
+# The LB cilium does not connect to the kube-apiserver. For now we use Kind
+# just to create Docker-in-Docker containers.
+kind create cluster --config kind-config.yaml
+cat cilium-lb.yaml | sed "s#{{IMG}}#$IMG#g" | kubectl apply -f -
```
please use helm instead of keeping `cilium-lb.yaml`

brb replied:
Helm installs bunch of stuff which is not needed (e.g. operator, CNI, etc). Is it possible to say to helm to install just cilium DS and cilium-config CM?

brb replied:
Also, we don't expect users to run the standalone LB via helm.

aanm replied:
> Is it possible to say to helm to install just cilium DS and cilium-config CM?

yes.

brb replied:
~@aanm Could you share the instructions? Thanks.~

brb replied:
@aanm Could we merge it as it is? In any case, I'm planning to rework our Helm template wrt loadbalancing. So then I could make it possible to reuse the template for these tests.

aanm replied:
Let's have the template now to avoid adding unnecessary files.

brb replied:
Done.

pchaigno commented on .github/workflows/l4lb.yaml:
```diff
@@ -0,0 +1,65 @@
+  setup-and-test:
+    name: Cilium L4LB XDP
+    # We need nested virtualisation which is supported only by MacOS runner
+    runs-on: macos-10.15
```
So if I understand correctly, this workflow is running K8s containers inside Kind containers inside a Fedora VM inside a macOS VM on an Azure server?

brb replied:
Yep. The complexity arises from the fact that there is no GH Action runner which supports cgroupv2(-only). Unfortunately, it seems that there is no plan to add such a runner: https://github.com/actions/virtual-environments/issues/2307. Alternatively, we could host a GH Action runner by ourselves, but AFAIK it's not recommended for OSS projects due to security concerns.

Another complication is that there is the 5 concurrent MacOS runners limit per account: https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration#usage-limits.

pchaigno replied:
> Another complication is that there is the 5 concurrent MacOS runners limit per account: https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration#usage-limits.

That might be okay if we run only on trigger comments. Worst case, we remove it from `test-me-please` and have it run only on demand and on schedule.

pchaigno commented on .github/workflows/l4lb.yaml:
```diff
@@ -0,0 +1,65 @@
+      - master
+    paths-ignore:
+      - 'Documentation/**'
+      - 'test/**'
```
Could we run this workflow only on trigger phrases as with the GKE/AKS/etc. workflows? I'm a bit worried of the impact it will have on our queue of GitHub jobs :scream_cat:

pchaigno commented on .github/workflows/l4lb.yaml:
```diff
@@ -0,0 +1,65 @@
+    runs-on: macos-10.15
+    timeout-minutes: 30
+    strategy:
+      fail-fast: false
```
What's the purpose of this? I think it only works with matrices: https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategyfail-fast.

pchaigno commented on .github/workflows/l4lb.yaml:
```diff
@@ -0,0 +1,65 @@
+        run: |
+          ln -sf ./test/l4lb-xdp/Vagrantfile ./Vagrantfile
+          # Retry if it fails (download.fedoraproject.org returns 404 sometimes)
+          # Spend up to 10 seconds on this
```
Could we copy the image to our usual Vagrant box location to avoid this? This looks like the sort of thing that will be either flaky or a pain for our runtime.

brb replied:
I'm afraid that fetching from the usual location will be more flaky, as we might hit the S3 rate-limiting issue (the reason why we have the VM image cache). So far I haven't seen the 404. Please see https://github.com/containerd/containerd/pull/4999 for more details.

pchaigno commented on .github/workflows/l4lb.yaml:
```diff
@@ -0,0 +1,65 @@
+            echo ::set-output name=tag::${{ github.sha }}
+          fi
+
+
```
Nit: unnecessary newline?

pchaigno commented on test/l4lb-xdp/test.sh:
```diff
@@ -0,0 +1,70 @@
+# complicate network setup, we will attach the following program which
+# terminates the tunnel.
+# The program is taken from the Linux kernel selftests.
+clang -O2 -Wall -target bpf -c test_tc_tunnel.c -o test_tc_tunnel.o
```
If I understand correctly, these two programs are both hacks that facilitate testing but differ from a typical user environment, right? What's the risk that we miss a regression in some way because of that?

brb replied:
> these two programs are both hacks that facilitate testing but differ from a typical user environment

Yep.

The `test_tc_tunnel` program is used to terminate the IPIP tunnel on a backend node. Alternatively, we could have created an IPIP tunnel dev. But that approach would be more cumbersome, as we would need to allocate an IP for it from a different subnet, make that subnet routable, etc. I think the chosen approach can be even legit for prod envs.

Meanwhile, the `bpf_xdp_veth_host` is a hack to overcome the veth XDP limitation. I doubt that any user would run XDP on veth (maybe `XDP_REDIRECT`, but that's not relevant to the standalone LB case).

> What's the risk that we miss a regression in some way because of that?

I don't see any big risk. The introduced GH Action is testing the standalone LB capabilities in bpf_host on XDP. The program execution does not depend on any XDP implementation details specific to a driver (minus the hack which does almost nothing). We could have chosen the full-blown Vagrant setup with VMs, but I don't see what additional value it could bring except for an increased CI resource usage and a longer test wait time. Catching any regression in an XDP driver code is not in the scope of these tests.

pchaigno replied:
Sounds good to me. Thanks for explaining!

pchaigno commented on test/l4lb-xdp/test_tc_tunnel.c:
```diff
@@ -0,0 +1,541 @@
+// SPDX-License-Identifier: GPL-2.0
```
Do we need to indicate this was copied from kernel sources?

brb replied:
Hmm, good question. Maybe let me add an explicit comment.

pchaigno reviewed (changes requested):
This feels fairly hackish but I understand why we need all that and don't have a cleaner/simpler solution :neutral_face: 

A couple comments below, but nothing really blocking. I would strongly prefer if we trigger this new workflow only on comments though. Our queue of GitHub jobs is already getting out of control when all contributors are online and sending PRs...

pchaigno commented on .github/workflows/l4lb.yaml:
```diff
@@ -0,0 +1,63 @@
+  ###
+
+concurrency:
+  group: "${{ github.workflow }} ${{ github.event.issue.pull_request.url || 'scheduled' }} ${{ github.event.comment.body }}"
```
That last part (`github.event.comment.body`) will generate errors when triggered by large comments. See https://github.com/cilium/cilium/commit/6440f293d27719d76d239ba12e3dd91914cf5548.

brb replied:
Thanks for writing details git commit messages. It helped to understand the reasoning and the issues behind the concurrency groups.

pchaigno commented on .github/workflows/l4lb.yaml:
```diff
@@ -0,0 +1,63 @@
+      - created
+  # Run every 6 hours
+  schedule:
+    - cron:  '0 2/6 * * *'
```
I noticed that [the schedule for other workflows](https://github.com/cilium/cilium/search?q=%22run+every+6+hours%22) is always off by one, such that no two workflows run at the same time. I'm not sure what's the rationale for that (cc @michi-covalent, cf. https://github.com/cilium/cilium/commit/3b0d4fbde75340a57273f1e79f661fcaf5df5c81).

brb replied:
Made it off by one (`0 4/6 * * *`).

brb replied:
> I'm not sure what's the rationale for that

Maybe to reduce a likelihood to hit the concurrent runners per account quota .

pchaigno replied:
Ah yes. That could be a good reason :+1:

brb committed:
helm: Add loadBalancer.{standalone,dsrDispatch}

This commit introduces the following Helm options:
- "loadBalancer.standalone" to enable the standalone Cilium L4LB.
- "loadBalancer.dsrDispatch" to choose the DSR dispatch mode.

Also, this commit replaces --node-port-{mode,acceleration} with
--bpf-lb-{mode,acceleration}, respectively. The former two were
deprecated by d73c572fcc.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on .github/workflows/l4lb.yaml:
```diff
@@ -0,0 +1,69 @@
+    timeout-minutes: 30
+
+    steps:
+      - uses: actions/checkout@v2
```
needs pinned commit like other actions have

aanm commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -538,15 +538,22 @@ data:
   bpf-lb-algorithm: {{ .Values.loadBalancer.algorithm | quote }}
 {{- end }}
 {{- if hasKey .Values.loadBalancer "acceleration" }}
-  node-port-acceleration: {{ .Values.loadBalancer.acceleration | quote }}
```
Are we deprecating this option? if yes it needs to be added in the upgrade notes

brb replied:
No, please see https://github.com/cilium/cilium/commit/d73c572fcc970bcb19bbcb86373dcc3595ec00f7 linked in the git commit msg.

aanm commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -538,15 +538,22 @@ data:
+  datapath-mode: lb-only
+{{- end }}
 {{- if hasKey .Values.loadBalancer "mode" }}
-  node-port-mode: {{ .Values.loadBalancer.mode | quote }}
```
Are we deprecating this option? if yes it needs to be added in the upgrade notes

brb replied:
Same as ^^.

aanm commented on .github/workflows/l4lb.yaml:
```diff
@@ -0,0 +1,69 @@
+name: Cilium L4LB XDP
```
Also, the filename should be `test-l4lb.yaml` to be consistent with other tests

brb replied:
I don't see any `test-*.yaml`:
```
$~/sandbox/gopath/src/github.com/cilium/cilium/.github/workflows [08:41:20]
> ls -a
.                   codeql-analysis.yml        go-check.yaml                       images.yaml        smoke-test-ipv6.yaml
..                  cyclonus-netpol-test.yaml  images-legacy-base-releases.yaml    kind-1.19.yaml     smoke-test.yaml
aks.yaml            docs.yaml                  images-legacy-hotfix-releases.yaml  l4lb.yaml
bpf-checks.yaml     eks.yaml                   images-legacy-releases.yaml         multicluster.yaml
build_commits.yaml  gke.yaml                   images-legacy.yaml                  nightly.yaml
```

aanm replied:
rebase and you shall see

errordeveloper commented on test/l4lb-xdp/Vagrantfile:
```diff
@@ -0,0 +1,34 @@
+# Vagrant box for testing kind with cgroup v2
```
Is Vagrant the only way to accomplish this? I can see some info on this in the PR description, but it would be quite helpful to have in a comment.

brb replied:
I've added a comment which hopefully helps to explain why we need this.

nebril commented on .github/workflows/tests-l4lb.yaml:
```diff
@@ -0,0 +1,69 @@
+  group: |
+    "${{ github.workflow }} ${{ github.event.issue.pull_request.url || 'scheduled' }}
+     ${{ (startsWith(github.event.comment.body, 'ci-l4lb') ||
+          startsWith(github.event.comment.body, 'test-me-please')) && 'trigger-phrase' }}"
```
Do we plan on having this test required for merging PRs? If that's the case, I expect that it will become a serious bottleneck for merging PRs, since number of macos runners is very limited.

Can we run this only for changes that touch load balancer (alternatively, we can have a skipping mechanism akin to bpf lint workflow: https://github.com/cilium/cilium/blob/master/.github/workflows/lint-bpf-checks.yaml).

errordeveloper replied:
> ...since number of macos runners is very limited

Oh, I've not noticed this! To this is even a greater reason to run this elsewhere, if cgroupsv2 is not available anywhere out of the box, an EKS setup with a custom node AMI should be quite doable.

errordeveloper replied:
Actually, I'd check Bottlerocket, it might have cgroupsv2, you can use it on EKS instead of AL2.

nebril replied:
I think it's 5 concurrent macos runners: https://docs.github.com/en/actions/reference/usage-limits-billing-and-administration

pchaigno replied:
I don't think a limit of 5 concurrent jobs will be an issue. The job currently takes less than 15min. So we would need to run 20 end-to-end tests per hour (19 if you take the scheduled workflow into account) just to reach the limit. We are not yet that busy a project :smile:

nebril replied:
Didn't check how long were the runs. 15 mins should be fine, but I would definitely keep an open mind about moving this somewhere else.

errordeveloper replied:
Limits are only one end of the stick. Apple doesn't allow to run macOS on any random VMs on any host machine, it has to run on Apple hardware. The hardware is going to be in much more limited supply and is more costly. The usage limits guard against too high of a demand, but they don't guarantee availability.

nebril reviewed (commented):
Helm changes look good, but I am concerned with being constrained by availability os macos executors. Plz let me know if we checked this somehow.

nbusseneau commented on .github/workflows/tests-l4lb.yaml:
```diff
@@ -0,0 +1,69 @@
+  cancel-in-progress: true
+
+jobs:
+  setup-and-test:
```
Job is missing an `if` condition, so by default it will run on every comment create as per the events defined at the top. If the intent is to use a specific trigger phrase (I think it is, considering the concurrency group above :D), we should guard with something like this:

```
    if: |
      (github.event.issue.pull_request && (
        startsWith(github.event.comment.body, 'ci-l4lb') ||
        (startsWith(github.event.comment.body, 'test-me-please'))
      )) ||
      (github.event_name == 'schedule' && github.repository == 'cilium/cilium') ||
      github.event.label.name == 'ci-run/l4lb'
```

nbusseneau replied:
@brb BTW, and I recommend we do so, if you do add a label filter `ci-run/l4lb` as proposed in the code snippet above, you may test the workflow prior to merging by uncommenting the following lines in a separate commit/testing PR:

```
  # pull_request:
  #  types:
  #    - "labeled"
```

and then adding the `ci-run/l4lb` to the testing PR. This will trigger the `labeled` event as per the filter, and we can test that the workflow behaves as intended (mostly... since it's a different event, we can't test everything) before merging for real :)

The `pull_request` event should be left commented in the real PR / real code to merge, because we don't want this to be possible for external contributors.

brb replied:
Thanks!

nbusseneau commented on .github/workflows/tests-l4lb.yaml:
```diff
@@ -0,0 +1,69 @@
+      - created
+  # Run every 6 hours
+  schedule:
+    - cron:  '0 4/6 * * *'
```
[Nit] The 4th slot is already used by the multicluster workflow, suggesting 5th to balance out actions a little bit:
```suggestion
    - cron:  '0 5/6 * * *'
```

nbusseneau reviewed (changes requested):
LGTM overall, I think we missed one thing re: how to trigger though.

christarazi reviewed (approved):
LGTM, let's add the codeowner changes to this PR as well.

kaworu commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -538,15 +538,22 @@ data:
   enable-auto-protect-node-port-range: {{ .Values.nodePort.autoProtectPortRange | quote }}
 {{- end }}
 {{- if hasKey .Values "loadBalancer" }}
+{{- if .Values.loadBalancer.standalone }}
```
Should we have this value added as a comment into `values.yml`?

brb replied:
Thanks, added!

kaworu commented on install/kubernetes/cilium/templates/cilium-configmap.yaml:
```diff
@@ -538,15 +538,22 @@ data:
-  node-port-acceleration: {{ .Values.loadBalancer.acceleration | quote }}
+  bpf-lb-acceleration: {{ .Values.loadBalancer.acceleration | quote }}
+{{- end }}
+{{- if hasKey .Values.loadBalancer "dsrDispatch" }}
+  bpf-lb-dsr-dispatch: {{ .Values.loadBalancer.dsrDispatch | quote }}
```
Ditto

nbusseneau reviewed (approved):
Thanks, the changes LGTM. I see you've tested with `pull_request` uncommented and the label: https://github.com/cilium/cilium/actions/runs/905705472 👍🏻

christarazi commented on CODEOWNERS:
```diff
@@ -252,6 +252,8 @@ Jenkinsfile.nightly @cilium/ci-structure
 /test/runtime/kafka.go @cilium/proxy @cilium/ci-structure
 /test/runtime/memcache.go @cilium/proxy @cilium/ci-structure
+# Standalone L4LB tests
+/test/l4lb @cilium/loadbalancer
```
All other test directories / files also include `@cilium/ci-structure`
